### PR TITLE
Add support for Python 3.14

### LINKED ISSUE #5472: Add support for python 3.14
## Description
Python 3.14 has been out since oct 2025 and python 3.15 should be coming later in the year. This ticket is to add support for 3.14 in the `kedro` main repo.

Related to https://github.com/kedro-org/kedro/pull/5206

## Context
https://devguide.python.org/versions/

### PR DESCRIPTION
**Pull request by lrcouto**
## Description

Adds Python 3.14 support to Kedro.

The most significant change here is that Python 3.14 changed the default multiprocessing start method on Linux from `fork` to `forkserver`. 

With `fork`, new worker processes are cloned directly from the parent. They inherit its full memory state, including any Kedro project configuration and open file descriptors. With `forkserver`, a server process is pre-spawned before any test or application code runs, and workers are forked from that server on demand. In this case, the server has none of the parent's runtime state. 

This required changes in two places. First, `Task._run_node_synchronization` only bootstrapped worker processes (configure_project, hook registration) when the start method was `spawn`. `forkserver` workers arrived unconfigured, so hooks were never registered and no hook callbacks fired. Second, `pytest-cov` does subprocesses by injecting a sitecustomize.py that imports and starts coverage in every child process. This works under `fork` (the child inherits a valid working directory) and under `spawn` (a fresh interpreter with a normal environment), but it crashes under `forkserver`.
 
When the server process runs sitecustomize.py, coverage calls os.path.realpath('.') during initialization and raises FileNotFoundError because the forkserver has no accessible working directory. The server crashes, and every test that creates a SyncManager or ProcessPoolExecutor gets an EOFError on the broken communication pipe. The fix was to bootstrap guard in task.py to include "forkserver" and force the start method to "spawn" in tests/conftest.py on Python 3.14+ (coverage handles spawn correctly, and it matches the behavior already validated on Windows and macOS). Also, updating the SKIP_ON_WINDOWS_AND_MACOS marker to also skip the fork-dependent tests when `forkserver` is active; and add "forkserver" as a valid KEDRO_MP_CONTEXT value so users can opt into it.

Worth noticing that pytest-cov [dropped support for multiprocessing 3 major releases ago](https://pytest-cov.readthedocs.io/en/latest/changelog.html#id36), so some jankiness is to be expected.

Also took the  opportunity to change some leftover text that still claimed that Kedro supports Python 3.9.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: RELEASE.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Upcoming Release
 
 ## Major features and improvements
+* * Add official support for Python 3.14.
```
Comment (ElenaKhaustova) [inline RELEASE.md]: **Critical:** Double bullet here — `* *` should be `*`. This will render incorrectly in markdown.

```suggestion
* Add official support for Python 3.14.
```

File: kedro/validation/type_extractor.py [hunk 2]
```diff
@@ -19,6 +19,18 @@
...
 
+def _type_name(tp: type) -> str:
+    """Return a human-readable name for a type, using str() for union types.
+
+    In Python 3.14, types.UnionType gained __name__ == "Union", so
+    getattr(tp, "__name__", str(tp)) would return "Union" instead of
+    the full representation like "list[str] | int".
+    """
+    if isinstance(tp, types.UnionType) or get_origin(tp) is Union:
+        return str(tp)
```
Comment (ElenaKhaustova) [inline kedro/validation/type_extractor.py:30]: **Suggestion:** Nice helper — the docstring explaining the Python 3.14 `UnionType.__name__` change is valuable context.

Consider also applying `_type_name` to the remaining `getattr(type_arg, "__name__", str(type_arg))` call in the union warning message (~line 145). If a type argument of a union is itself a `UnionType` (e.g. `(int | str) | float`), it would hit the same 3.14 regression. Edge-case-y, but would be consistent and defensive.

Also, while the existing `test_conflicting_types_getattr_fallback` test covers this path, a small dedicated test for `_type_name` would make the intent explicit.

File: tests/runner/test_parallel_runner.py [hunk 3]
```diff
@@ -456,7 +456,7 @@ def test_get_executor_default(self, mocker):
         assert isinstance(executor, ProcessPoolExecutor)
         assert hasattr(executor, "_mp_context")
 
-    @pytest.mark.parametrize("context", ["fork", "spawn"])
+    @pytest.mark.parametrize("context", ["fork", "spawn", "forkserver"])
```
Comment (ElenaKhaustova) [inline tests/runner/test_parallel_runner.py:459]: **Suggestion:** Now that `"forkserver"` is parametrized here, the skip reason on line 462 (`"fork context is not available on Windows"`) is slightly misleading — `forkserver` is also unavailable on Windows. Consider updating to something like `"fork and forkserver contexts are not available on Windows"`. Minor nit.

**Review by ElenaKhaustova — COMMENTED** [review]
## Kedro PR Review
> Generated with `review-kedro-pr` skill.

### Overview
- **Critical:** 1 | **Suggestions:** 3

### Critical (must fix before merge)
- [ ] `RELEASE.md:L4` — Double bullet `* *` formatting error

### Suggestions (consider improving)
- [ ] `kedro/validation/type_extractor.py:L143` — Apply `_type_name` consistently in the union warning message
- [ ] `tests/runner/test_parallel_runner.py:L462` — Skip reason is misleading now that `forkserver` is parametrized
- [ ] `kedro/validation/type_extractor.py:L22` — Consider a dedicated unit test for `_type_name`

### Notes
- **General review:** The PR is well-scoped, focused, and the description is excellent — it thoroughly explains the `fork` vs `forkserver` vs `spawn` problem and the rationale for each change. The approach of forcing `spawn` in tests to work around the `pytest-cov`/`forkserver` incompatibility is pragmatic and well-documented.
- **PR checklist:** The RELEASE.md entry is present but has a formatting error (see Critical). Docs are updated. CI matrix is updated. The `install.md` fix (`3.9+` → `3.10+`) is a good cleanup. None of the checklist items are checked off.
- **Architecture:** No violations. `kedro.runner.task` importing from `kedro.framework.project` is covered by the allowed exception. No new cross-layer imports.
- **API usage:** The extension of `_run_node_synchronization` to bootstrap on `forkserver` is correct and follows the existing pattern. Adding `"forkserver"` as a valid `KEDRO_MP_CONTEXT` value is a clean, backward-compatible addition.

**Comment by lrcouto** [comment]
Applied the suggested changes.

**Review by lrcouto — COMMENTED** [review]
## Kedro PR Review
> Generated with `review-kedro-pr` skill.

### Overview
- **Critical:** 0 | **Suggestions:** 2

### Critical (must fix before merge)
No issues found.

### Suggestions (consider improving)
- [ ] `tests/framework/session/test_session_extension_hooks.py:L36` — `SKIP_ON_WINDOWS_AND_MACOS` no longer accurately describes the skip condition (also skips on Python 3.14+ Linux); consider renaming.
- [ ] `kedro/runner/task.py:L133` — The `"forkserver"` bootstrap branch is not directly test-covered; `tests/conftest.py` forces `spawn` on 3.14+.

### Notes
- **General review:** Clean, well-scoped PR with an excellent description that correctly diagnoses the `forkserver` / pytest-cov incompatibility. Doc updates for `run_a_pipeline.md` are thorough. The `Python 3.9+ → 3.10+` correction in `install.md` is a fair bundled cleanup (flagged in the PR description).
- **PR checklist:** RELEASE.md entry present and well-formatted. Docs updated. No breaking changes. No Viz impact.
- **Architecture:** No layer violations. All new imports are stdlib (`multiprocessing`, `types`, `typing`).
- **API usage:** No new datasets/runners/hooks/config-loaders. Existing `AbstractRunner` and `Task` contracts preserved.

File: tests/framework/session/test_session_extension_hooks.py [hunk 4]
```diff
@@ -30,11 +30,12 @@
...
 
-# Window and MacOS(Python>3.7) default with `spawn` process which reload modules
-# and cause test fails
+# Windows and macOS use `spawn` by default; Python 3.14+ Linux uses `forkserver`.
+# Both methods don't inherit file descriptors, so log records from subprocesses
+# can't reach the test's QueueHandler, causing hook count assertions to fail.
 SKIP_ON_WINDOWS_AND_MACOS = pytest.mark.skipif(
-    multiprocessing.get_start_method() == "spawn",
-    reason="Due to bug in parallel runner",
+    multiprocessing.get_start_method() in ("spawn", "forkserver"),
```
Comment (lrcouto) [inline tests/framework/session/test_session_extension_hooks.py:37]: **Suggestion:** With `"forkserver"` added to this skipif, the constant name `SKIP_ON_WINDOWS_AND_MACOS` (line 36) is no longer accurate — it now also skips on Python 3.14+ Linux, where `forkserver` is the default. Consider renaming to something like `SKIP_WHEN_NON_FORK_START_METHOD` for clarity. There are 7 usages, all in this file, so the rename is contained.

File: kedro/runner/task.py [hunk 5]
```diff
@@ -129,7 +129,10 @@ def _run_node_synchronization(
             The node argument.
 
         """
-        if multiprocessing.get_start_method() == "spawn" and package_name:
+        if (
+            multiprocessing.get_start_method() in ("spawn", "forkserver")
```
Comment (lrcouto) [inline kedro/runner/task.py:133]: **Suggestion:** The newly added `"forkserver"` branch is not directly exercised by any test — `tests/conftest.py` forces `spawn` on Python 3.14+, so CI never runs this code path with `forkserver` as the actual start method. A small unit test that patches `multiprocessing.get_start_method` to return `"forkserver"` and asserts `_bootstrap_subprocess` is called would guard against future regressions if someone edits the membership set.

File: tests/framework/session/test_session_extension_hooks.py [hunk 6]
```diff
@@ -30,11 +30,12 @@
...
 
-# Window and MacOS(Python>3.7) default with `spawn` process which reload modules
-# and cause test fails
-SKIP_ON_WINDOWS_AND_MACOS = pytest.mark.skipif(
-    multiprocessing.get_start_method() == "spawn",
-    reason="Due to bug in parallel runner",
+# Windows and macOS use `spawn` by default; Python 3.14+ Linux uses `forkserver`.
+# Both methods don't inherit file descriptors, so log records from subprocesses
+# can't reach the test's QueueHandler, causing hook count assertions to fail.
+SKIP_WHEN_NON_FORK_START_METHOD = pytest.mark.skipif(
```
Comment (merelcht) [inline tests/framework/session/test_session_extension_hooks.py:36]: It sounds like at some point (when 3.14 and above are the only supported versions) we might not have any way of testing this anymore. Is there any way in which we can test this going forward that doesn't require skipping on certain versions?

**Review by merelcht — APPROVED** [review]
LGTM! 👍 

With one question around the test skipping going forward.

File: tests/framework/session/test_session_extension_hooks.py (hunk 6 above)
Comment (lrcouto) [reply tests/framework/session/test_session_extension_hooks.py:36]: We could refactor the tests to force them to use fork. The issue that I see with it is that it would no longer represent the behavior users would encounter when running Kedro on Python 3.14+, which uses forkserver by default, so it'd be kinda pointless.

We could also exclude the tests from coverage, since pytest-cov is what's causing the issue. It's also not ideal.

The cleanest solution would be figuring out a way to get coverage to work with forkserver. E2E tests pass so I'm assuming the features that require parallel running themselves are working on 3.14 as expected, but I'd also like to be able to have functioning unit tests.

**Review by ElenaKhaustova — APPROVED** [review]
LGTM 👍

**Review by ankatiyar — APPROVED** [review]
